modules/web-console/web-console-server/src/test/java/org/apache/ignite/console/web/security/AuthenticationTest.java

```diff
@@ -0,0 +1,324 @@
+
+        GridTestUtils.assertThrows(null, () -> {
+            authMgr.authenticate(new UsernamePasswordAuthenticationToken(
+                USER_EMAIL,
+                "wrong_password"
+            ));
+
+            return null;
+        }, BadCredentialsException.class, messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials"));
+
+        Account acc = accountsRepo.getByEmail(USER_EMAIL);
+
+        assertEquals(2, acc.getFailedLoginAttempts());
+        assertNotEquals(0L, acc.getLastFailedLogin());
+
+        Thread.sleep((long)Math.pow(accCfg.getAuthentication().getInterval(), Math.log(acc.getFailedLoginAttempts())));
+
+        authMgr.authenticate(new UsernamePasswordAuthenticationToken(
+            USER_EMAIL,
+            "password"
+        ));
+
+        acc = accountsRepo.getByEmail(USER_EMAIL);
+
+        assertEquals(0, acc.getFailedLoginAttempts());
+        assertEquals(0L, acc.getLastFailedLogin());
+    }
+
+    /**
+     * Perform login attempt with wrong credentials
+     * Immediately try to login with wrong credentials
+     * Immediately try to login with wrong credentials
+     * Verify the User is revoked
+     * Verify the error and error message about timeout, no. of failed attempts is 3  and lastFailedLogin is updated
+     * Immediately try to login with valid credentials
+     * Verify the error, no. of failed attempts and lastFailedLogin
+     * Wait for cooldown timeout
+     * Try to login with valid credentials
+     * Verify the error, no. of failed attempts and lastFailedLogin
+     */
+    @Test
+    public void shouldLockWithTooManyAttempts() throws InterruptedException {
+        createTestUser();
+
+        GridTestUtils.assertThrows(null, () -> {
+            authMgr.authenticate(new UsernamePasswordAuthenticationToken(
+                USER_EMAIL,
+                "wrong_password"
+            ));
+
+            return null;
+        }, BadCredentialsException.class, messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials"));
+
+        GridTestUtils.assertThrows(null, () -> {
+            authMgr.authenticate(new UsernamePasswordAuthenticationToken(
+                USER_EMAIL,
+                "wrong_password"
+            ));
+
+            return null;
+        }, BadCredentialsException.class, messages.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials"));
+
+        GridTestUtils.assertThrows(null, () -> {
+            authMgr.authenticate(new UsernamePasswordAuthenticationToken(
+                USER_EMAIL,
+                "password"
+            ));
+
+            return null;
+        }, LockedException.class, "Account is currently locked. Try again later");
+
+        Account acc = accountsRepo.getByEmail(USER_EMAIL);
+
+        assertEquals(3, acc.getFailedLoginAttempts());
+        assertNotEquals(0L, acc.getLastFailedLogin());
```

Comment:
Would be better to check that `getLastFailedLogin()` is not just differs from 0, but that it changes - so it is X after 1st fail, Y after 2nd fail and so on

Reply:
Done